warp10/src/main/java/io/warp10/script/functions/MFILTER.java

```diff
@@ -0,0 +1,92 @@
+//
+//   Copyright 2021  SenX S.A.S.
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+
+package io.warp10.script.functions;
+
+import io.warp10.script.NamedWarpScriptFunction;
+import io.warp10.script.WarpScriptException;
+import io.warp10.script.WarpScriptStack;
+import io.warp10.script.WarpScriptStack.Macro;
+import io.warp10.script.WarpScriptStackFunction;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+public class MFILTER extends NamedWarpScriptFunction implements WarpScriptStackFunction {
+
+  public MFILTER(String name) {
+    super(name);
+  }
+
+  @Override
+  public Object apply(WarpScriptStack stack) throws WarpScriptException {
+
+    Object top = stack.pop();
+
+    boolean pushIndex = true;
+    if (top instanceof Boolean) {
+      pushIndex = (Boolean) top;
+      top = stack.pop();
+    }
+    
+    if (!(top instanceof Macro)) {
+      throw new WarpScriptException(getName() + " expects a macro on top of the stack.");
+    }
+
+    Macro macro = (Macro) top;
+    
+    top = stack.pop();
+    
+    if (!(top instanceof Map)) {
+      throw new WarpScriptException(getName() + " expects a MAP below the macro on top of the stack.");
+    }
+
+    Map<?, ?> map = (Map) top;
+
+    LinkedHashMap<Object, Object> result = new LinkedHashMap<Object, Object>(map.size());
```

Comment:
Allocating an identical number of elements may be inefficient because we do not know ahead of time the proportion of elements which will be retained.